### PR TITLE
Fix weights not properly initialized due to shape mismatch

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Currently, if there is some weight shape mismatched between the model and the checkpoint, and if ignore_mismatched_sizes=True, that/those weight(s) won't get initialized by the model's `_init_weights` method, and could get crazy values like 1e37.

This could make the training gets `nan loss value` from the beginning, (then `Trainer` will change this to `0.0`) and the training won't have any progress (loss always 0.0).

One example is by running `src/transformers/modeling_utils.py` (add `ignore_mismatched_sizes=True`).

We usually set `ignore_mismatched_sizes=True` when we want to perform classification tasks using an existing model but to another task having different number of targets.

This PR aims to fix this issue.

### CODE REVIEW

**Comment by ydshieh** [comment]
I will explain it a bit more tomorrow.

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -3957,13 +3957,14 @@ def _fix_key(key):
...
         # retrieve unintialized modules and initialize before maybe overriding that with the pretrained weights.
         if _fast_init:
-            if remove_prefix_from_model:
-                _loaded_keys = [f"{prefix}.{k}" for k in loaded_keys]
-            elif add_prefix_to_model:
-                _loaded_keys = [k[len(prefix) + 1 :] for k in loaded_keys]
-            else:
-                _loaded_keys = loaded_keys
-            set_initialized_submodules(model, _loaded_keys)
+            if not ignore_mismatched_sizes:
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:3960]: If `ignore_mismatched_sizes` is True, it means there is some shape mismatch of weight(s) between a model and a checkpoint.

Currently, in this case, those model weights won't be initialized by `_init_weights`:
  - because we defaults to `_fast_init` in `from_pretrained`
  - and this looks the weight names in the checkpoint, marked them as already initialized, and won't apply `_init_weights` to those weights.
  - The above action doesn't take into account if there is a shape mismatch (as this information is done after where we deal with `_fast_init`)

In order to fix the issue, this PR just not to use `_fast_init` if `ignore_mismatched_sizes` is `True`.

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -3957,13 +3957,14 @@ def _fix_key(key):
...
+                if remove_prefix_from_model:
+                    _loaded_keys = [f"{prefix}.{k}" for k in loaded_keys]
+                elif add_prefix_to_model:
+                    _loaded_keys = [k[len(prefix) + 1 :] for k in loaded_keys]
+                else:
+                    _loaded_keys = loaded_keys
+                set_initialized_submodules(model, _loaded_keys)
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:3967]: This is to mark weights being already initialized if they are in the checkpoint. Don't perform this if `ignore_mismatched_sizes=True`

File: tests/test_modeling_common.py [hunk 3]
```diff
@@ -2889,6 +2889,42 @@ def test_load_with_mismatched_shapes(self):
...
+                        )
+                    self.assertIn("the shapes did not match", cl.out)
+
+                    for name, param in new_model.named_parameters():
+                        if param.requires_grad:
+                            self.assertIn(
+                                ((param.data.mean() * 1e9).round() / 1e9).item(),
+                                [0.0, 1.0],
+                                msg=f"Parameter {name} of model {model_class} seems not properly initialized",
+                            )
```
Comment (ydshieh) [inline tests/test_modeling_common.py:2926]: A combination of `test_initialization` and `test_load_with_mismatched_shapes` to check the initialization is done correctly when there is shape mismatch

File: tests/test_modeling_common.py [hunk 4]
```diff
@@ -2889,6 +2889,42 @@ def test_load_with_mismatched_shapes(self):
                     else:
                         new_model_without_prefix(input_ids)
 
+    def test_mismatched_shapes_have_properly_initialized_weights(self):
+        if not self.test_mismatched_shapes:
+            return
```
Comment (amyeroberts) [inline tests/test_modeling_common.py:2894]: Can we return with a skip message here?

**Review by amyeroberts — COMMENTED** [review]
Thanks for digging into this and fixing! 

Only request before merging is to add a test to make sure weights of the expected shape are correctly loaded even if `ignore_mismatched_sizes=True` i.e. only the weight matrices that need to be resized are randomly initialized.

**Comment by ydshieh** [comment]
> only the weight matrices that need to be resized are randomly initialized.

That is a very good point, and I did think of it once. Here we have keys in checkpoint and keys in the model's `state_dcit()`.
They are not always identical, as could be seen [here](https://github.com/huggingface/transformers/blob/5aec50ecaf9c1c039cde85881f0586110f845859/src/transformers/modeling_utils.py#L4007-L4021).

Those conversions are only inside `modeling_utils.py` and not exposed at all. So I don't have the way to do the suggested testing.

In most cases, I would say they are identical. I can still implement it, but I am a bit worried we also have edge cases to skip.

WDYT?
--------------------------------------------------

But the change in this PR add initialization before any actual weight loading. So for those weights intended to be loaded,  they are loaded as before this PR. So if there is any problem, it must be something wrong before this PR.

**Comment by amyeroberts** [comment]
@ydshieh Would it be possible to create a small dummy model with e.g. two linear layers - one we set to explicit values and the other we resize to test this? So we don't have to try to handle complex layer naming in the state dict?

**Comment by ydshieh** [comment]
> @ydshieh Would it be possible to create a small dummy model with e.g. two linear layers - one we set to explicit values and the other we resize to test this? So we don't have to try to handle complex layer naming in the state dict?

Yes, that could be done, similar to [this](https://github.com/huggingface/transformers/blob/main/tests/test_modeling_common.py#L434-L436)

I will do it. But TBH, the above test and the new one to be added are model class independent - and we should probably move them into a new test class rather than being inside `ModelTesterMixin`.

Let me update this PR with your suggestion, and leave the moving of test methods in a follow up PR.

**Comment by amyeroberts** [comment]
SGTM - thanks!

**Comment by ydshieh** [comment]
A new test `def test_matched_shapes_have_loaded_weights_when_some_mismatched_shapes_exist(self)` is added

**Review by amyeroberts — APPROVED** [review]
Very nice - thanks for digging into this, resolving and adding comprehensive tests!
